### PR TITLE
Clear staging after commit

### PR DESCRIPTION
**Pull request by shinhs0506**
- **PR Description**
fixes https://github.com/jesseduffield/lazygit/issues/2363
invoke stage panel refresh on successful commit

- **Please check if the PR fulfills these requirements**

* [x] Cheatsheets are up-to-date (run `go run scripts/cheatsheet/main.go generate`)
* [x] Code has been formatted (see [here](https://github.com/jesseduffield/lazygit/blob/master/CONTRIBUTING.md#code-formatting))
* [x] Tests have been added/updated (see [here](https://github.com/jesseduffield/lazygit/blob/master/pkg/integration/README.md) for the integration test guide)
* [ ] Text is internationalised (see [here](https://github.com/jesseduffield/lazygit/blob/master/CONTRIBUTING.md#internationalisation))
* [x] Docs (specifically `docs/Config.md`) have been updated if necessary
* [x] You've read through your own file changes for silly mistakes etc

### CODE REVIEW

**Comment by mark2185** [comment]
LGTM

File: pkg/integration/tests/commit/staged.go
Comment (jesseduffield) [inline pkg/integration/tests/commit/staged.go:58]: we should add an assertion here that the staging panel has been refreshed (there are examples of asserting on a view's contents in various tests so I'd take a a look at those)

**Review by jesseduffield — CHANGES_REQUESTED** [review]
one thing :) (haven't functionally tested)

File: pkg/integration/tests/commit/staged.go
Comment (jesseduffield) [inline pkg/integration/tests/commit/staged.go:56]: The Tap function is only really needed when we want to make an assertion about another view and then afterwards chain more method calls onto the original one. But in this case it's not necessary; we can instead go:
```go
t.Views().StagingSecondary().
	IsEmpty()
	
t.Views().Staging().
	IsFocused().
	Content(Contains("+myfile content")).
	Content(DoesNotContain("+with a second line"))

```

File: pkg/integration/tests/commit/staged_without_hooks.go
Comment (jesseduffield) [inline pkg/integration/tests/commit/staged_without_hooks.go:55]: same here

**Review by jesseduffield — CHANGES_REQUESTED** [review]
one thing :)

**Comment by shinhs0506** [comment]
resolved :)

**Comment by jesseduffield** [comment]
Sweet. Code looks good, I wanna manually test this before merging (hopefully today)

**Comment by jesseduffield** [comment]
Testing this locally I notice a flicker in the main panel when we commit from either the main panel or the files panel. We wanna find a way around that before we merge this

**Comment by shinhs0506** [comment]
are you able to upload a recording of the flicker? just so that we are on the same page

**Comment by jesseduffield** [comment]
In this recording I first open up the master branch (that's what lg does) and then show there's no flickering, and then on this PR's branch (that's what gg does) there is flickering:
https://user-images.githubusercontent.com/8456633/216798926-35f40573-1490-46cd-a2a2-62be536bc443.mov

**Comment by stefanhaller** [comment]
Since I run into this bug quite often, and it's confusing and annoying every time, I wonder if it would be an option to merge the fix in the current state and open an issue to address the flicker problem later. What do you think @jesseduffield?

(BTW, I find the flickering barely noticeable anyway, on my machine at least.)

**Comment by stefanhaller** [comment]
I debugged the flicker issue, here's what I found:

- onCommitSuccess calls Refresh(Staging) after the commit was made, but before the files panel is refreshed
- refreshStagingPanel calls handleStagingEscape because both main views have become empty
- this calls PushContext(gui.State.Contexts.Files) (but that context is already active in this case)
- this deactivates the files context, and then activates it again
- this calls RenderToMainViews
- since the files panel hasn't been refreshed yet, this calls the same diff command again, for the same file that had the staged changes before. Since the commit has just been made, this diff is now empty and the screen goes black.
- Then the files panel is refreshed, and the main view displays whatever file is selected now.

The flicker can be avoided by not deactivating and activating the context if the currently active context is pushed again. Here's a commit that does this: https://github.com/stefanhaller/lazygit/commit/3b7b171adf7f33c7b6ed0dd016e26ae

@jesseduffield Let me know what you think about this approach.

@shinhs0506 Can you try cherry-picking the above commit to your branch to see if it improves the flicker? (It does for me.)

**Comment by jesseduffield** [comment]
@stefanhaller that commit looks sensible to me. But I'll need to do some testing on it: it may be that we have code that depends on the activateContext function being called on already-active contexts. That is: there may be too kinds of things happening in that function: one for 'activated the first time' and another for 'activated at any time'.

**Comment by jesseduffield** [comment]
@stefanhaller one thing I notice upon testing is that the options view (at the bottom left in blue) does not appear when you initially start the program

**Comment by stefanhaller** [comment]
5b10b333d843 should fix the missing options view problem on startup (if that's the right way to do it). Can you think of any other problems with this approach?

**Comment by jesseduffield** [comment]
I do worry that calling `PushContext` before calling `gui.g.MainLoop()` could get us into trouble, but if the integration tests all pass I'm happy

**Comment by jesseduffield** [comment]
I've done some testing and it looks good. I'm happy to push those commits to this branch. Did you want to squash them first @stefanhaller ?

**Comment by stefanhaller** [comment]
Great. I'm fine with merging the branch as is, with the two commits cherry-picked to it.

**Comment by stefanhaller** [comment]
@jesseduffield Is this still waiting for something, or did you just forget to merge it? 😉

**Pull request merged**
